Realm/Tests/MigrationTests.mm

```diff
@@ -652,6 +653,33 @@ - (void)testMigrationBlockCalledWhenSchemaVersionHasChanged {
     XCTAssertTrue(migrationCalled);
 }
 
+#pragma mark - Async Migration
+
+- (void)testAsyncMigration {
+    RLMRealmConfiguration *c = [RLMRealmConfiguration new];
+    c.schemaVersion = 1;
+    @autoreleasepool { XCTAssertNoThrow([RLMRealm realmWithConfiguration:c error:nil]); }
+    XCTAssertNil(RLMGetAnyCachedRealmForPath(c.pathOnDisk.UTF8String));
+    XCTestExpectation *ex = [self expectationWithDescription:@"async-migration"];
+    __block bool migrationCalled = false;
+    c.schemaVersion = 2;
+    c.migrationBlock = ^(__unused RLMMigration *migration, __unused uint64_t oldSchemaVersion) {
+        migrationCalled = true;
+    };
+    [RLMRealm asyncOpenWithConfiguration:c
+                           callbackQueue:dispatch_get_main_queue()
+                                 callback:^(RLMRealm * _Nullable realm, NSError * _Nullable error) {
+        [ex fulfill];
```

Comment:
Fulfilling the expectation should always be the last thing done in the block when the block isn't running on the queue owned by `RLMTestCase`, as otherwise the following assertions in the block could be checked after the test case has completed and been marked as passing (there's a bunch of tests in this PR that do this).